### ISSUE TITLE
Crouch and Sprint not recognizing Input Map

### ISSUE THREAD

**Issue by Phlegmlee** [ColormaticStudios/quality-godot-first-person-2#18]
I created Input Maps (project settings) for all key bindings, all work fine with the exception of crouch and sprint. The string values I originally chose were "CROUCH" and "SPRINT". I attempted changing the string value thinking it might be conflicting with something, but any string value returns the error. I made sure my spellings matched up as well.

With the code:
```
@export var CROUCH : String = "CROUCH"
@export var SPRINT : String = "SPRINT"
```

I receive the push error that comes default to the FPC add-on:
```
E 0:00:00:0304   character.gd:117 @ check_controls(): No control mapped for crouch. Please add an input map control. Disabling crouching.
  <C++ Source>   core/variant/variant_utility.cpp:1091 @ push_error()
  <Stack Trace>  character.gd:117 @ check_controls()
                 character.gd:95 @ _ready()
```

I have traced what I believe is all the crouching and sprinting code without finding the root cause, if I suppress the error from disabling crouching and sprinting respectively I get repeating errors from the following processes:
```
E 0:00:02:0741   character.gd:250 @ handle_state(): The InputMap action "crouch" doesn't exist.
  <C++ Error>    Condition "!InputMap::get_singleton()->has_action(p_action)" is true. Returning: false
  <C++ Source>   core/input/input.cpp:292 @ is_action_just_pressed()
  <Stack Trace>  character.gd:250 @ handle_state()
                 character.gd:162 @ _physics_process()

E 0:00:02:0757   character.gd:231 @ handle_state(): The InputMap action "sprint" doesn't exist.
  <C++ Error>    Condition "!InputMap::get_singleton()->has_action(p_action)" is true. Returning: false
  <C++ Source>   core/input/input.cpp:287 @ is_action_pressed()
  <Stack Trace>  character.gd:231 @ handle_state()
                 character.gd:162 @ _physics_process()
```

I am running Godot 4.2 Stable.

**Comment by Phlegmlee** [comment]
<img width="863" alt="image" src="https://github.com/ColormaticStudios/quality-godot-first-person-2/assets/152333980/b6a4713a-1855-467a-9e7b-c9c2068cce14">

**Comment by Phlegmlee** [comment]
Confirmed that rolling back to a previous version, one that doesn't do states, solves the issue. I do not know exactly what version I have on my old project but looking at the version history it seems to be the previous release.

**Comment by CZakarya** [comment]
With the error line:
```
character.gd:250 @ handle_state(): The InputMap action "crouch" doesn't exist.
```
It seems like either the FPC or the engine is getting `"crouch"` as a lowercase value, while you have `CROUCH` in the input map.

From a first glance, I can't see what the error is (I know, crazy) but I have a few things for you to try.
- Change the InputMap names to lowercase ("crouch", "sprint") (just for debug, if this fixes it we need to find the underlying bug)
- Try to recreate this bug with the latest release in a new project
- If you changed any of the code in the FPC, either try reverting those lines in a separate branch of your project, or inspect those changes closely

You mentioned reverting to an older version that doesn't use states. This project has always used a state machine, ever since version one. It started out with a more primitive one though. The new state machine (26f7e773e2a322c42416113a9ed9cd1a490cf3ef) slightly changed how the state controls were handled.

A possible cause is that I accidentally hardcoded the string `"crouch"` into the `Input.is_action_pressed` function in that commit on line 127, 138, 149, and 155, but @ItWinVis pointed this out in issue #4 and I fixed it in (01ff47493c715fd925a45fd566a7297d2fb4d62a) and (a1729c1cd554112254573b3dd6554dbbb61f66c4).

**Comment by Phlegmlee** [comment]
Good news, but also a bit of bad news, using lowercase in the InputMap clears up the errors and it now switches states appropriately. This works with the string being lowercase or uppercase in the script. (Super weird)

The only code I changed according to my history is uncommenting the full controller support code and changing the string names to my own InputMap names instead of the default UI ones.

I have attached a zip file with the recreated error in its fixed state using lowercase InputMap strings and uppercase strings in the script. I started by using uppercase for both and it gave the same error.

In the recreation I left all other code alone.
[testing_fpc.zip](https://github.com/user-attachments/files/16060184/testing_fpc.zip)

Below is the code mentioned that you fixed and I confirm that I do have the updated, fixed version for all instances of the function call.
```
if Input.is_action_pressed(SPRINT) and state != "crouching":
if Input.is_action_pressed(CROUCH) and state != "sprinting":

And all others...
```

Based on what I am finding it may be a bug on Godot's side with specifically the words crouch and sprint not being recognized while uppercase in InputMap, all other strings are upper case and they work fine for me.

**Comment by Phlegmlee** [comment]
Here is a screenshot of all my Input Maps and a code snipped for the script.
```
@export_group("Controls")
@export var JUMP : String = "JUMP"
@export var LEFT : String = "LEFT"
@export var RIGHT : String = "RIGHT"
@export var FORWARD : String = "FORWARD"
@export var BACKWARD : String = "BACKWARD"
@export var PAUSE : String = "PAUSE"
@export var CROUCH : String = "CROUCH"
@export var SPRINT : String = "SPRINT"
```
<img width="863" alt="image" src="https://github.com/ColormaticStudios/quality-godot-first-person-2/assets/152333980/4b2c97be-4543-4333-a0dd-eaae89c0bef9">
Everything works as it should with this.

**Comment by CZakarya** [comment]
I tested the project with all the inputs uppercased and sure enough, it had an issue with just the sprint and crouch inputs.

There are two possibilities that I can see for this:
- If it is a problem in the FPC, it would probably be in the `handle_state` function since that's where both of them are handled and they are done quite similarly. I've combed through this function multiple times and there is no hardcoding nor any alternate button detection systems. If this is the case, it has to be really well hidden.
- If it is a problem with the engine, then the InputMap code is doing something with the sprint and crouch inputs, or it is somehow saving the original values of the export variable and not accepting the changed code.

I know there is either a bug in the physics engine or a feature of InputMap where if a crouch button is pressed, you cannot fall off of a platform. Just try to walk off of the main test platform in the FPC, it won't let you. I didn't add any code to the FPC to do this, this is Godot Engine itself. This also happened in QFPC V1.

I'll probably have to go through the Godot engine code for InputMap to see if it handles sprint and crouch inputs differently and maybe do a bug report or a PR.

**Comment by Phlegmlee** [comment]
I checked the stack to see what the variables CROUCH and SPRINT were initializing as and the values that I input have no effect, they always init as lowercase crouch and sprint. I don't know where it would be as I have checked all the files in the add-on, but, somewhere it is hard coded. Unless there is another way to get this to occur.

<img width="924" alt="image" src="https://github.com/ColormaticStudios/quality-godot-first-person-2/assets/152333980/e717983e-8ef8-4897-8bc1-5f0fd02115cc">

I checked the Input.cpp and Input_Map.cpp files for Godot located here: https://github.com/godotengine/godot/tree/master/core/input

I found that the error code from the engine is directly from the exact_match method but there is nothing to do with crouch and sprint respectively.
